File: boat-sample/src/test/java/sample/xyz/srclab/common/collect/CollectSample.java
package sample.xyz.srclab.common.collect;

import org.testng.annotations.Test;
import xyz.srclab.common.base.Nums;
import xyz.srclab.common.collect.ArrayCollects;
import xyz.srclab.common.collect.ListOps;
import xyz.srclab.common.test.TestLogger;

import java.util.ArrayList;
import java.util.List;

public class CollectSample {

    private static final TestLogger logger = TestLogger.DEFAULT;

    @Test
    public void testArray() {
        String[] strings = ArrayCollects.newArray("1", "2", "3");
        ArrayCollects.asList(strings).set(0, "111");
        //111
        logger.log("string[0]: {}", strings[0]);
    }

    @Test
    public void testCollect() {
        List<String> list = new ArrayList<>();
        list.add("1");
        list.add("2");
        list.add("3");
        ListOps<String> listOps = ListOps.opsFor(list);
        int sum = listOps.addAll(ArrayCollects.newArray("4", "5", "6"))
                .removeFirst()
                .map(it -> it + "0")
                .map(Nums::toInt)
                .reduce(Integer::sum);
        //200
        logger.log("sum: {}", sum);
    }
}

File: boat-sample/src/test/kotlin/sample/kotlin/xyz/srclab/common/collect/CollectSample.kt
package sample.kotlin.xyz.srclab.common.collect

import org.testng.annotations.Test
import xyz.srclab.common.collect.ListMap.Companion.toListMap
import xyz.srclab.common.collect.MutableListMap.Companion.toMutableListMap
import xyz.srclab.common.collect.MutableSetMap.Companion.toMutableSetMap
import xyz.srclab.common.collect.SetMap.Companion.toSetMap
import xyz.srclab.common.collect.addElements
import xyz.srclab.common.test.TestLogger
import java.util.*

class CollectSample {

    @Test
    fun testMultiMap() {
        val setMap = mapOf("s" to setOf("1", "2", "3")).toSetMap()
        //setMap: {s=[1, 2, 3]}
        logger.log("setMap: {}", setMap)

        val mutableSetMap = mutableMapOf("s" to mutableSetOf("1", "2", "3")).toMutableSetMap()
        mutableSetMap.add("s", "9")
        mutableSetMap.addAll("s", LinkedHashSet<String>().addElements("11", "12", "13"))
        //mutableSetMap: {s=[1, 2, 3, 9, 11, 12, 13]}
        logger.log("mutableSetMap: {}", mutableSetMap)

        val listMap = mapOf("s" to listOf("1", "2", "3")).toListMap()
        //listMap: {s=[1, 2, 3]}
        logger.log("listMap: {}", listMap)

        val mutableListMap = mutableMapOf("s" to mutableListOf("1", "2", "3")).toMutableListMap()
        mutableListMap.add("s", "9")
        mutableListMap.addAll("s", LinkedList<String>().addElements("11", "12", "13"))
        //mutableListMap: {s=[1, 2, 3, 9, 11, 12, 13]}
        logger.log("mutableListMap: {}", mutableListMap)
    }

    companion object {
        private val logger = TestLogger.DEFAULT
    }
}